File: src/services/auth/AuthService.ts
import type { User, LoginCredentials, AuthTokens } from '@/types';
import type { AuthApi } from '@/services/api';
import { TokenManager } from './TokenManager';
import { STORAGE_KEYS } from '@/utils';

export interface IAuthService {
  login(email: string, password: string): Promise<AuthTokens>;
  logout(): Promise<void>;
  refreshToken(): Promise<AuthTokens>;
  getCurrentUser(): User | null;
  isAuthenticated(): boolean;
}

export class AuthService implements IAuthService {
  private tokenManager: TokenManager;
  private api: AuthApi;
  private currentUser: User | null = null;

  constructor(tokenManager: TokenManager, api: AuthApi) {
    this.tokenManager = tokenManager;
    this.api = api;
    this.loadCurrentUser();
  }

  async login(email: string, password: string): Promise<AuthTokens> {
    try {
      const credentials: LoginCredentials = { email, password };
      const response = await this.api.login(credentials);
      
      this.tokenManager.setTokens(response.tokens);
      this.currentUser = response.user;
      
      localStorage.setItem(STORAGE_KEYS.USER_ID, response.user.id);
      localStorage.setItem(STORAGE_KEYS.USER_EMAIL, response.user.email);
      
      return response.tokens;
    } catch (error) {
      console.error('Login failed:', error);
      throw new Error('登录失败，请检查用户名和密码');
    }
  }

  async logout(): Promise<void> {
    try {
      await this.api.logout();
    } catch (error) {
      console.error('Logout API failed:', error);
    } finally {
      this.tokenManager.clearTokens();
      this.currentUser = null;
      
      localStorage.removeItem(STORAGE_KEYS.USER_ID);
      localStorage.removeItem(STORAGE_KEYS.USER_EMAIL);
    }
  }

  async refreshToken(): Promise<AuthTokens> {
    try {
      const tokens = await this.api.refreshToken();
      this.tokenManager.setTokens(tokens);
      return tokens;
    } catch (error) {
      console.error('Token refresh failed:', error);
      await this.logout();
      throw new Error('Token刷新失败，请重新登录');
    }
  }

  getCurrentUser(): User | null {
    return this.currentUser;
  }

  isAuthenticated(): boolean {
    return this.tokenManager.getAccessToken() !== null && !this.tokenManager.isTokenExpired();
  }

  private async loadCurrentUser(): Promise<void> {
    const token = this.tokenManager.getAccessToken();
    if (!token) {
      return;
    }

    try {
      const userId = localStorage.getItem(STORAGE_KEYS.USER_ID);
      const userEmail = localStorage.getItem(STORAGE_KEYS.USER_EMAIL);
      
      if (userId && userEmail) {
        this.currentUser = {
          id: userId,
          email: userEmail,
        };
      }
    } catch (error) {
      console.error('Failed to load current user:', error);
      this.currentUser = null;
    }
  }
}

File: flutter_im_client/lib/services/auth_service.dart
import 'package:logging/logging.dart';
import '../models/auth.dart';
import '../models/user.dart';
import '../constants/api_constants.dart';
import '../utils/exceptions.dart';
import 'http_client_service.dart';
import 'secure_storage_service.dart';

class AuthService {
  static final AuthService _instance = AuthService._internal();
  factory AuthService() => _instance;
  AuthService._internal();

  final HttpClientService _httpClient = HttpClientService();
  final SecureStorageService _storage = SecureStorageService();
  final Logger _logger = Logger('AuthService');

  User? _currentUser;

  User? get currentUser => _currentUser;
  bool get isAuthenticated => _currentUser != null;

  Future<void> init() async {
    await _httpClient.init();
  }

  Future<AuthResponse> login(String email, String password) async {
    try {
      _logger.info('Attempting login for email: $email');
      
      await _httpClient.init();
      final loginRequest = LoginRequest(email: email, password: password);
      try {
        final response = await _httpClient.post<Map<String, dynamic>>(
          ApiConstants.login,
          data: loginRequest.toJson(),
        );
        final authResponse = AuthResponse.fromJson(response.data!);
        await _storage.saveToken(authResponse.token);
        await _storage.saveRefreshToken(authResponse.refreshToken);
        _currentUser = authResponse.user;
        _logger.info('Login successful for user: ${authResponse.user.email}');
        return authResponse;
      } on NetworkException catch (e) {
        // Fallback to mock login for backend without auth endpoints or connection failures
        if (e.statusCode == null || e.statusCode == 0 || e.statusCode == 404 || e.statusCode == 501) {
          _logger.info('Auth endpoint unavailable, using mock login');
          final now = DateTime.now().millisecondsSinceEpoch;
          String? userId;
          final lower = email.toLowerCase();
          if (lower == 'test1@example.com') userId = 'user1';
          if (lower == 'test2@example.com') userId = 'user2';
          if (lower == 'test3@example.com') userId = 'user3';
          if (userId == null) {
            throw AuthException(message: 'Unsupported account for mock login', type: AuthErrorType.invalidCredentials);
          }
          final token = 'mock_jwt_${userId}_${now}';
          await _storage.saveToken(token);
          await _storage.saveRefreshToken('mock_refresh_${userId}_${now}');
          final me = await _httpClient.get<Map<String, dynamic>>(ApiConstants.me);
          _currentUser = User.fromJson(me.data!);
          return AuthResponse(token: token, refreshToken: 'mock_refresh_${userId}_${now}', user: _currentUser!);
        }
        rethrow;
      }
    } on NetworkException catch (e) {
      _logger.severe('Login failed: ${e.message}');
      throw AuthException(
        message: e.message,
        type: e.statusCode == 401 
            ? AuthErrorType.invalidCredentials 
            : AuthErrorType.networkError,
      );
    } catch (e) {
      _logger.severe('Login error: $e');
      throw AuthException(
        message: 'Login failed: ${e.toString()}',
        type: AuthErrorType.unknown,
      );
    }
  }

  Future<void> logout() async {
    try {
      _logger.info('Logging out user: ${_currentUser?.email}');
      
      final token = await _storage.getToken();
      if (token != null) {
        await _httpClient.post(ApiConstants.logout);
      }
    } catch (e) {
      _logger.warning('Logout API call failed: $e');
    } finally {
      // Clear local data regardless of API call result
      await _storage.clearAll();
      _currentUser = null;
      _logger.info('Logout completed');
    }
  }

  Future<bool> checkAuthentication() async {
    try {
      final token = await _storage.getToken();
      if (token == null) {
        return false;
      }

      // Try to get current user info
      final response = await _httpClient.get<Map<String, dynamic>>(ApiConstants.me);
      _currentUser = User.fromJson(response.data!);
      
      return true;
    } catch (e) {
      _logger.warning('Authentication check failed: $e');
      return false;
    }
  }

  Future<String?> getAccessToken() async {
    return await _storage.getToken();
  }

  Future<void> refreshToken() async {
    try {
      await _httpClient.refreshToken();
    } catch (e) {
      _logger.severe('Token refresh failed: $e');
      throw AuthException(
        message: 'Token refresh failed',
        type: AuthErrorType.tokenExpired,
      );
    }
  }

  Future<User> getCurrentUser() async {
    if (_currentUser != null) {
      return _currentUser!;
    }

    try {
      final response = await _httpClient.get<Map<String, dynamic>>(ApiConstants.me);
      _currentUser = User.fromJson(response.data!);
      return _currentUser!;
    } catch (e) {
      _logger.severe('Failed to get current user: $e');
      throw AuthException(
        message: 'Failed to get current user',
        type: AuthErrorType.unknown,
      );
    }
  }
}
